### PR TITLE
KVL-810 Report the correct gRPC error code on failures due to concurrent transactions

### PR DESCRIPTION
**Pull request by hubert-da**
### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala [hunk 1]
```diff
@@ -268,10 +269,14 @@ private[apiserver] final class ApiSubmissionService private[services] (
...
   private def toStatus(errorCause: ErrorCause) =
     errorCause match {
-      case e: ErrorCause.DamlLf =>
-        Status.INVALID_ARGUMENT.withDescription(e.explain)
-      case e: ErrorCause.LedgerTime =>
-        Status.ABORTED.withDescription(e.explain)
+      case cause @ ErrorCause.DamlLf(error) =>
+        error match {
+          case ContractNotFound(_) | ReplayMismatch(_) =>
+            Status.ABORTED.withDescription(cause.explain)
```
Comment (hubert-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala:275]: Don't we actually want to return the NOT_FOUND code?

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/keys/KeyMonotonicityValidation.scala [hunk 2]
```diff
@@ -41,7 +41,7 @@ private[keys] object KeyMonotonicityValidation {
         recordTime,
         transactionCommitter.buildRejectionLogEntry(
           transactionEntry,
-          RejectionReason.Inconsistent("Causal monotonicity violated"),
+          RejectionReason.InvalidLedgerTime("Causal monotonicity violated"),
```
Comment (hubert-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/keys/KeyMonotonicityValidation.scala:44]: @rautenrieth-da do you think that's fine?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala [hunk 3]
```diff
@@ -60,8 +60,8 @@ final class ContractKeysIT extends LedgerTestSuite {
       assertGrpcError(fetchFailure, Status.Code.INVALID_ARGUMENT, "couldn't find key")
       assertGrpcError(
         lookupByKeyFailure,
-        Status.Code.INVALID_ARGUMENT,
-        Some(Pattern.compile("Disputed|Inconsistent")),
+        Status.Code.ABORTED,
```
Comment (hubert-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala:63]: Do you think we need more tests?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala (hunk 1 above)
Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala:275]: Not sure; I find it suspect to return `NOT_FOUND` for something that's a command, not a query.

File: ledger/participant-integration-api/src/main/scala/platform/store/CompletionFromTransaction.scala [hunk 4]
```diff
@@ -36,30 +37,23 @@ private[platform] object CompletionFromTransaction {
...
         Code.ABORTED
-      case _: RejectionReason.SubmitterCannotActViaParticipant =>
-        Code.PERMISSION_DENIED
+      }
+      case _: SubmitterCannotActViaParticipant => Code.PERMISSION_DENIED
     }
   }
 
-  private def toParticipantRejection(reason: domain.RejectionReason): RejectionReason =
+  def toParticipantRejection(reason: domain.RejectionReason): RejectionReason =
```
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/main/scala/platform/store/CompletionFromTransaction.scala:48]: Please move this somewhere else if you're going to reuse it.

File: ledger/participant-integration-api/src/main/scala/platform/store/CompletionFromTransaction.scala [hunk 5]
```diff
@@ -36,30 +37,23 @@ private[platform] object CompletionFromTransaction {
   // * -Xfatal-warnings _MUST_ be enabled
   def toErrorCode(rejection: RejectionReason): Code = {
```
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/main/scala/platform/store/CompletionFromTransaction.scala:38]: This too.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/keys/KeyMonotonicityValidation.scala (hunk 2 above)
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/keys/KeyMonotonicityValidation.scala:44]: Is that tested somewhere? I would have expected to see a corresponding test change.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala [hunk 6]
```diff
@@ -177,7 +177,7 @@ final class ContractKeysIT extends LedgerTestSuite {
     } yield {
       assertGrpcError(
         duplicateKeyFailure,
-        Status.Code.INVALID_ARGUMENT,
+        Status.Code.ABORTED,
         Some(Pattern.compile("DuplicateKey|Inconsistent")),
```
Comment (fabiotudone-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala:181]: Can this pattern be refined as well, i.e., which is it?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala [hunk 7]
```diff
@@ -268,10 +269,14 @@ private[apiserver] final class ApiSubmissionService private[services] (
 
   private def toStatus(errorCause: ErrorCause) =
     errorCause match {
-      case e: ErrorCause.DamlLf =>
-        Status.INVALID_ARGUMENT.withDescription(e.explain)
-      case e: ErrorCause.LedgerTime =>
-        Status.ABORTED.withDescription(e.explain)
+      case cause @ ErrorCause.DamlLf(error) =>
+        error match {
+          case ContractNotFound(_) | ReplayMismatch(_) =>
```
Comment (fabiotudone-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala:274]: Why was `ReplayMismatch` absent? Is it tested?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala (hunk 6 above)
Comment (hubert-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractKeysIT.scala:181]: Not sure why I missed it, thanks

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala (hunk 7 above)
Comment (hubert-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala:274]: Samir suggested putting it here. Will add a test

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/keys/KeyMonotonicityValidation.scala (hunk 2 above)
Comment (hubert-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/keys/KeyMonotonicityValidation.scala:44]: Good point, I will add a test if Robert is fine with the change

Comment (rautenrieth-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/keys/KeyMonotonicityValidation.scala:44]: `InvalidLedgerTime` is the correct rejection reason here.

I expect that this will change the error visible in the ledger API from `INVALID_ARGUMENT` to `ABORTED` (see [here](https://github.com/digital-asset/daml/blob/main/ledger/participant-integration-api/src/main/scala/platform/store/CompletionFromTransaction.scala#L42)), which is also the better error message (given the [respective descriptions](https://grpc.github.io/grpc/core/md_doc_statuscodes.html)).

File: ledger/participant-integration-api/src/main/scala/platform/store/CompletionFromTransaction.scala (hunk 4 above)
Comment (hubert-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/CompletionFromTransaction.scala]: Moved to `Conversions`

**Comment by bame-da** [comment]
This change is fine with me. I'd classify it as a bugfix as it brings error codes in line with spec. My only request would be that the documentation on the protos is updated in line with this change:

- https://github.com/digital-asset/daml/blob/main/ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_submission_service.proto#L36-L43
- https://github.com/digital-asset/daml/blob/main/ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_service.proto
- Whatever other services are affected

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_service.proto [hunk 8]
```diff
@@ -28,6 +28,7 @@ service CommandService {
   // - ``INVALID_ARGUMENT``: if the payload is malformed or is missing required fields
   // - ``RESOURCE_EXHAUSTED``: if the number of in-flight commands reached the maximum (if a limit is configured)
   // - ``UNAVAILABLE``: if the participant is not yet ready to submit commands or if the service has been shut down.
+  // - ``ABORTED``: if a key is missing or duplicated due to for example contention on resources
```
Comment (hubert-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_service.proto:31]: Does anyone have any idea how to verify if I'm not missing any other service like the active contracts or transaction service?

**Comment by hubert-da** [comment]
Please take a look if the commit message is ok.

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_service.proto (hunk 8 above)
Comment (fabiotudone-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_service.proto:31]: Not a quick one but the only one I can think about: scanning through them with the help of the docs and pondering if this fix is applicable or not.

**Comment by cocreature** [comment]
> - [Integration Kit] Fix gRPC status codes for inconsistency rejections and DamlLf errors

Can we include what we changed? "fix" is quite vague.

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_service.proto [hunk 9]
```diff
@@ -39,6 +40,7 @@ service CommandService {
...
```
Comment (miklos-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_service.proto:43]: I'd suggest being more specific here regarding the type of the key:
```suggestion
  // - ``ABORTED``: if a contract key is missing or duplicated due to for example contention on resources
```

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_service.proto [hunk 10]
```diff
@@ -39,6 +40,7 @@ service CommandService {
   // - ``INVALID_ARGUMENT``: if the payload is malformed or is missing required fields
```
Comment (miklos-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_service.proto:40]: What's the ordering of these error codes? They don't seem to line up with the numerical order of the [gRPC error codes](https://grpc.github.io/grpc/core/md_doc_statuscodes.html). I don't expect to align them as part of this PR.

File: ledger/ledger-api-test-tool-on-canton/BUILD.bazel [hunk 11]
```diff
@@ -74,11 +74,13 @@ conformance_test(
...
-        # The following three contract key tests require uniqueness and are tested in conformance-test-contract-keys below
+        # The following three contract key tests require uniqueness
         "--exclude=ContractKeysIT,ContractKeysIT:CKFetchOrLookup,ContractKeysIT:CKNoFetchUndisclosed,ContractKeysIT:CKMaintainerScoped" +
         ",ParticipantPruningIT" +  # see "conformance-test-participant-pruning" below
         ",ConfigManagementServiceIT,LedgerConfigurationServiceIT" +  # dynamic config management not supported by Canton
-        ",ClosedWorldIT",  # Canton currently fails this test with a different error (missing namespace in "unallocated" party id)
+        ",ClosedWorldIT" +  # Canton currently fails this test with a different error (missing namespace in "unallocated" party id)
+        # Some of gRPC status codes changed from INVALID_ARGUMENT to ABORTED (see https://github.com/digital-asset/daml/pull/9218):
+        ",SemanticTests:SemanticDoubleSpendBasic,SemanticTests:SemanticDoubleSpendShared,SemanticTests:SemanticPrivacyProjections,SemanticTests:SemanticDivulgence",
```
Comment (miklos-da) [inline ledger/ledger-api-test-tool-on-canton/BUILD.bazel:83]: Where do we track the fixing of the excludes added here?

File: ledger/participant-integration-api/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker [hunk 12]
```diff
@@ -0,0 +1 @@
+mock-maker-inline
```
Comment (miklos-da) [inline ledger/participant-integration-api/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: Why is this needed?

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/keys/KeyMonotonicityValidationSpec.scala [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+  private def buildStateValue(activeAt: Instant) = {
+    DamlStateValue
+      .newBuilder()
+      .setContractKeyState(
+        DamlContractKeyState
+          .newBuilder()
+          .setActiveAt(Conversions.buildTimestamp(activeAt))
+      )
+      .build()
+  }
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/keys/KeyMonotonicityValidationSpec.scala:82]:
```suggestion
  private def buildStateValue(activeAt: Instant) =
    DamlStateValue
      .newBuilder()
      .setContractKeyState(
        DamlContractKeyState
          .newBuilder()
          .setActiveAt(Conversions.buildTimestamp(activeAt))
      )
      .build()
```

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/keys/KeyMonotonicityValidationSpec.scala [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    )
+  }
+
+  "checkContractKeysCausalMonotonicity" should {
+    "create StepContinue in case of correct keys" in {
+      KeyMonotonicityValidation.checkContractKeysCausalMonotonicity(
+        mock[TransactionCommitter],
+        None,
+        Set(testKey),
+        Map(testKey -> buildStateValue(baseTime.minusSeconds(1))),
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/keys/KeyMonotonicityValidationSpec.scala:44]: You may improve test readability by renaming the builder method to something like `aStateValueActiveAt` or similar, plus renaming `baseTime` to `ledgerEffectiveTime` or `ledgerTime`. This way, it would be obvious that activeness is being tweaked:
```scala
  ...
  Map(testKey -> aStateValueActiveAt(ledgerEffectiveTime.minusSeconds(1))),
  ...
```

File: ledger/participant-integration-api/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker (hunk 12 above)
Comment (hubert-da) [reply ledger/participant-integration-api/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: It allows to mock final classes

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_service.proto (hunk 10 above)
Comment (hubert-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_service.proto:40]: I contacted Kamil who documented them initially not that long ago. There is no particular order. I can change this if you'd like.

Comment (miklos-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_service.proto:40]: It's fine to change this as part of a different PR -- I'm OK with merging this one as-is.

**Review by miklos-da — APPROVED** [review]

File: ledger/ledger-api-test-tool-on-canton/BUILD.bazel (hunk 11 above)
Comment (cocreature) [reply ledger/ledger-api-test-tool-on-canton/BUILD.bazel:83]: No idea, my team is not responsible for either canton or the ledger api test tool :shrug:

Comment (cocreature) [reply ledger/ledger-api-test-tool-on-canton/BUILD.bazel:83]: Maybe @oliverse-da tracks this somewhere

Comment (oliverse-da) [reply ledger/ledger-api-test-tool-on-canton/BUILD.bazel:83]: Yes, I track these in canton. Let me take a look at these failing tests.

Comment (oliverse-da) [reply ledger/ledger-api-test-tool-on-canton/BUILD.bazel:83]: These tests are likely failing on canton because the canton build has embedded inside it an older version of the ledger api server, i.e. with the next canton release and upgrade in the daml repo, these test should pass again given that it is the ledger api server's `ApiSubmissionService` that now returns `ABORTED`.

**Comment by hubert-da** [comment]
@cocreature please let me know if you are fine with the changelog and the compatibility and Canton conformance test exclusions, thanks!

**Review by cocreature — APPROVED** [review]
Thank you!

**Review by andreaslochbihler-da — APPROVED** [review]
Since the error codes are currently determined by the Ledger API server, this should also work for Canton. However, independently of whether we use `ABORTED` or `INVALID_ARGUMENT`, this will not suffice to distinguish definite answers from concurrency problems, which is likely to be needed for a better command deduplication story. So looks reasonable for now, but we may have to refine this in the future.

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_service.proto (hunk 8 above)
Comment (hubert-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_service.proto]: I went through them (and the tests as well) and see no other service that is applicable, thanks.

File: ledger/ledger-api-test-tool-on-canton/BUILD.bazel (hunk 11 above)
Comment (hubert-da) [reply ledger/ledger-api-test-tool-on-canton/BUILD.bazel:84]: Also, created `KVL-876` as a follow-up.

Comment (oliverse-da) [reply ledger/ledger-api-test-tool-on-canton/BUILD.bazel:84]: Thanks @hubert-da

Comment (oliverse-da) [reply ledger/ledger-api-test-tool-on-canton/BUILD.bazel:84]: Follow-up happening here: https://github.com/digital-asset/daml/pull/9356
